### ISSUE TITLE
a<<32 bit rotations have issues in Clang when the "-O2" or "-O3" switch is used

### ISSUE THREAD

**Issue by samboy** [samboy/rg32hash#2]
Clang is broken when compiled with the "-O3" or "-O2" flag.  To wit:

```c
#include <stdio.h>
#include <stdint.h>
// Compile this with "clang -O3 debug.c -o debug" then run "./debug"
int main(){
        uint32_t n=1, j = 32;
        n <<= j;
        printf("%d\n",n); // Should return 0, doesn't in clang w/ -O3 flag
}
```

Now, as it turns out page 84 of http://www.open-std.org/jtc1/sc22/wg14/www/docs/n1124.pdf states that "If the value of the right [bitwise shift] operand is negative or is greater than or equal to the width of the promoted left operand, the behavior is undefined."

However, GCC and TCC do the right thing here (foo<<32 should be 0 if foo is a 32-bit unsigned int); why not Clang at higher optimization levels?

Here is what LLVM/Clang say on the subject: http://blog.llvm.org/2011/05/what-every-c-programmer-should-know.html

Now, because of how I used `a << 32`, it would do the right thing if either:

1) `a` remains unchanged (the x86 behavior, as per the LLVM blog) or
2) `a` becomes 0 (The powerPC behavior as per the LLVM blog)

(This is why I didn’t catch the bug until now)

However, for some reason, a becomes "15" at higher Clang optimizations in the above code.

**Comment by samboy** [comment]
I have updated nanorg32-vertical.c and tinyrg32.c to work around this bug. I still need to check the other C language implementations (nanorg32.c will need to be updated; others may also have the same issue).

**Comment by samboy** [comment]
I have updated all of the C implementations to not trigger the Clang a<<32 bug, since the C standard allows clang’s bug.  Fixed as of commit 492e9a8bcc0c52676e7e94fe371cd486178d23d3 _
